File: scripts/rollcheck.cpp
/*
 * Copyright (C) 2004-2026 Metaphonic Labs
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General
 * Public License along with this program; if not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*
 * rollcheck -- the piano roll's drawing, with no window around it.
 *
 *     scripts/rollcheck -p build/plugins/ gen/airports.gen
 *     scripts/rollcheck -p build/plugins/ --dump gen/airports.gen
 *
 * RollCanvas is the desktop's piano roll with the widget taken off
 * (src/RollCanvas.h), and the browser compiles the same file and draws
 * with it in the mirror worker. Two claims follow from that, and this
 * harness is both of them.
 *
 * The first is that the drawing says what the roll is for. Half of this
 * widget is the *scheduled future* -- what the piece has already decided
 * and not yet played -- and the page's old roll could not show it,
 * because a tape is what has been delivered. A list of ops is exactly
 * where that can be asserted without a screenshot: outlined rectangles,
 * to the right of the now-line, in a piece that is running.
 *
 * The second is that the two builds draw the same picture. `--dump'
 * prints the list the native class produced -- the ops and the strings
 * they index -- and wasm/web/rollcheck.mjs plays the same seeded piece to
 * the same frame through the module and diffs the two, the way
 * nodecheck.mjs diffs an edit against dspedit's. A seeded piece replays
 * identically (docs/GEN_FORMAT.md, gencheck), and the roll's drawing at a
 * given frame is a function of the piece and the frame, so anything left
 * over is the two builds disagreeing.
 *
 * Which is why this binary records rather than rasterises: it is linked
 * against cairo-canvas2d, the same cairo the browser module draws
 * through, instead of against real cairo. Nothing else in the native tree
 * is, and scripts/CMakeLists.txt says what that costs.
 *
 * The transport is stepped the way the module steps it -- once per
 * window, to the time that window's last frame falls on -- because the
 * comparison is only worth anything if both sides ask the roll to draw
 * the same instant of the same piece.
 *
 * Exit status is the number of failures.
 */

#include "config.h"

#include <stdio.h>
#include <string.h>

#include <filesystem>
#include <map>
#include <string>
#include <vector>

#include <glibmm.h>

#include "think.h"

#include "cairo2d.h"
#include "cairomm/context.h"

#include "libthink/thSynthCommand.h"
#include "thcPlugin.h"
#include "thcGenFile.h"
#include "thcScheduler.h"

#include "RollCanvas.h"

/* The window the browser's page opens its synth with, and the rate. The
   transport is stepped in these, so both sides reach the same instant by
   the same arithmetic rather than by two roundings of one number. */
static const int    WINDOW = 256;
static const int    RATE   = 48000;

/* How much of the piece is played before the roll is asked to draw, and
   how many frames it is then drawn for. Long enough that there is history
   behind the now-line and a queue in front of it; the frames matter
   because the pitch range eases toward its fit one draw at a time, so
   "the same frame" means the same number of pictures as well as the same
   instant. */
static const double SECONDS = 8.0;
static const int    FRAMES  = 30;

/* And how long the piece is played for the prune check at the end, which
   wants a transport well past four times the shortest span the roll can
   be set to -- 20s of scrub window, so 40s of playing leaves half the
   run behind the cutoff. Not part of the comparison, so it costs the
   diff nothing. */
static const double LONG_SECONDS = 40.0;

/* The size the comparison draws at. Not square, because the roll is a
   band: a shape it will never be asked for is a shape whose arithmetic
   nobody would notice being wrong. */
static const int    WIDTH  = 900;
static const int    HEIGHT = 180;

static int failures = 0;

static void
check (bool good, const std::string &what)
{
    printf("%s  %s\n", good ? "ok  " : "FAIL", what.c_str());

    if (!good)
        failures++;
}

/* Every composer module in the tree, by name. The same scan gencheck
   makes, and for the same reason: a piece names its stages by plugin
   name and nothing else can resolve one. */
static void
loadComposers (const std::string &pluginDir,
               std::map<std::string, thcPlugin *> &out)
{
    std::filesystem::path root =
        std::filesystem::path(pluginDir) / "composer";
    std::error_code ec;

    if (!std::filesystem::is_directory(root, ec))
        return;

    for (const auto &f : std::filesystem::directory_iterator(root, ec))
    {
        if (ec)
            break;

        if (f.path().extension() != PLUGIN_SUFFIX)
            continue;

        thcPlugin *p = new thcPlugin(f.path().string());

        if (p->state() != thcPlugin::LOADED)
        {
            delete p;
            continue;
        }

        out[p->name()] = p;
    }
}

/* ---- playing a piece the way the module plays one --------------------- */

/* The scheduler stepped to `seconds', one window at a time.
 *
 * thinkweb.cpp's step(): the target is the transport time the window's
 * last frame falls on, counted from the origin, and by nothing measured
 * -- so the piece is a function of the file and the seed. The same
 * sequence of targets here is what makes the two drawings comparable.
 *
 * The synth is drained every so often for gencheck's reason: every
 * delivered note posts a command for an audio thread that is not here,
 * and a full ring drops the SET_CHANNEL a piece's own instrument arrives
 * on. */
static void
play (thcScheduler &sched, thSynth &synth, double seconds)
{
    const double step = (double)WINDOW / RATE;
    long windows = 0;

    sched.start();

    for (long k = 1; sched.now() < seconds && sched.running(); k++)
    {
        sched.stepTransportTo(k * step);

        if (++windows % 64 == 0)
            synth.process();
    }

    synth.process();
}

/* ---- the list, out of cairo2d's recorder ------------------------------ */

/* Everything drawn since the context was made, as the three tables the
   page reads: no surfaces, because the roll blits nothing. */
struct List
{
    std::vector<float>       ops;
    std::vector<std::string> strings;
};

static List
listOf (cairo_t *cr)
{
    List out;
    const float *ops = cairo2d_ops(cr);
    const int words = cairo2d_op_words(cr);

    for (int i = 0; i < words; i++)
        out.ops.push_back(ops[i]);

    for (int i = 0; i < cairo2d_string_count(cr); i++)
        out.strings.push_back(cairo2d_string(cr, i));

    return out;
}

/* The roll drawn for `frames' frames, with the list from the last one.
 *
 * One recorder and one cairomm face over it, restarted per frame, which
 * is exactly what the module does (wasm/web/twdraw.cpp): the list is read
 * out before the next draw begins, on both sides. Two of them, or a fresh
 * context per frame, would be two arrangements being compared rather than
 * one drawing. */
static List
drawRoll (RollCanvas &roll, int width, int height, int frames)
{
    static cairo_t *cr = NULL;
    static Cairo::RefPtr<Cairo::Context> ctx;
    List last;

    if (cr == NULL)
    {
        cr = cairo2d_create();
        ctx = Cairo::Context::create(cr);
    }

    for (int i = 0; i < frames; i++)
    {
        cairo2d_begin(cr);
        roll.draw(ctx, width, height);
        last = listOf(cr);
    }

    return last;
}

/* The dump wasm/web/rollcheck.mjs reads: JSON, because the other side is
 * JavaScript, and %.9g because that round-trips a float exactly.
 *
 * The run's own numbers go out with it -- how long the piece was played
 * for, at what window and rate, how many frames it was drawn for and how
 * big -- so the other side drives itself from these rather than from a
 * second copy of them that could quietly stop matching. Every one of them
 * has to agree or the two lists are of two different instants. */
static void
dump (const List &list)
{
    printf("{\"seconds\":%g,\"frames\":%d,\"width\":%d,\"height\":%d,"
           "\"window\":%d,\"rate\":%d,",
           SECONDS, FRAMES, WIDTH, HEIGHT, WINDOW, RATE);
    printf("\"ops\":[");

    for (size_t i = 0; i < list.ops.size(); i++)
        printf("%s%.9g", i ? "," : "", (double)list.ops[i]);

    printf("],\"strings\":[");

    for (size_t i = 0; i < list.strings.size(); i++)
    {
        printf("%s\"", i ? "," : "");

        for (const char c : list.strings[i])
            if (c == '"' || c == '\\')
                printf("\\%c", c);
            else if ((unsigned char)c < 0x20)
                printf("\\u%04x", c);
            else
                printf("%c", c);

        printf("\"");
    }

    printf("]}\n");
}

/* ---- what the list has to have in it ---------------------------------- */

/* cairo2d's opcodes, as much of them as this needs. The arity table is
 * the package's and is walked here the way drawcheck.mjs walks it in
 * JavaScript: an op the table does not know, or a list that does not end
 * where the table says it ends, is a list the page would replay as
 * operands read as opcodes.
 */
static bool
walk (const std::vector<float> &ops, std::string &why, int &count)
{
    count = 0;

    for (size_t i = 0; i < ops.size(); )
    {
        const int op = (int)ops[i++];
        int arity = cairo2d_op_arity(op);

        if (cairo2d_op_name(op) == NULL)
        {
            why = "op " + std::to_string(op) + " at word " +
                  std::to_string(i - 1) + " is not one";
            return false;
        }

        if (arity < 0)                  /* SET_DASH says its own length */
        {
            if (i >= ops.size())
            {
                why = "a dash at word " + std::to_string(i - 1) +
                      " has no count";
                return false;
            }

            arity = (int)ops[i] + 2;
        }

        if (i + (size_t)arity > ops.size())
        {
            why = "op " + std::to_string(op) + " at word " +
                  std::to_string(i - 1) + " runs off the end";
            return false;
        }

        i += (size_t)arity;
        count++;
    }

    return true;
}

int
main (int argc, char *argv[])
{
    Glib::init();

    std::string pluginDir, genFile;
    bool dumping = false;

    for (int i = 1; i < argc; i++)
    {
        if (strcmp(argv[i], "-p") == 0 && i + 1 < argc)
            pluginDir = argv[++i];
        else if (strcmp(argv[i], "--dump") == 0)
            dumping = true;
        else
            genFile = argv[i];
    }

    if (pluginDir.empty() || genFile.empty())
    {
        fprintf(stderr,
                "usage: rollcheck -p <plugindir> [--dump] <file.gen>\n");
        return 2;
    }

    std::map<std::string, thcPlugin *> plugins;

    loadComposers(pluginDir, plugins);

    if (plugins.empty())
    {
        fprintf(stderr, "rollcheck: no composer modules in %s -- build "
                "the plugins first\n", pluginDir.c_str());
        return 2;
    }

    thSynth synth(pluginDir, WINDOW, RATE);
    thcScheduler sched(&synth);
    thcGenLoader loader(plugins);

    if (!loader.load(genFile, &sched))
    {
        fprintf(stderr, "rollcheck: %s did not load\n", genFile.c_str());

        for (size_t i = 0; i < loader.errors().size(); i++)
            fprintf(stderr, "  %s\n", loader.errors()[i].c_str());

        return 2;
    }

    /* Built after the load, as the mirror's is: what it keeps is what
       the scheduler delivers from here on, and the instrument
       applications a load itself delivers are not part of the piece. */
    RollCanvas roll(&sched);

    play(sched, synth, SECONDS);

    const List list = drawRoll(roll, WIDTH, HEIGHT, FRAMES);

    if (dumping)
    {
        dump(list);
        return 0;
    }

    /* ---- the list is one ---- */

    std::string why;
    int count = 0;
    const bool walked = walk(list.ops, why, count);

    check(walked,
          walked ? std::filesystem::path(genFile).filename().string() +
                   " at " + std::to_string((int)SECONDS) + "s is " +
                   std::to_string(count) + " ops in " +
                   std::to_string(list.ops.size()) + " words"
                 : why);

    /* ---- and the future is in it ----
     *
     * The assertion this whole extraction is for. A delivered note is a
     * filled rectangle and a scheduled one is an outline, so what says
     * the future half is being drawn is a stroked rectangle to the right
     * of the now-line -- and the now-line is at spanPast of spanPast plus
     * spanFuture across the width, which is the only place in the drawing
     * that ratio is spelled twice.
     *
     * Read out of the list rather than out of the canvas's own state,
     * because what is being checked is the *drawing*: a roll that kept a
     * perfect pending queue and drew none of it would pass every other
     * question here.
     */
    const double nowX = WIDTH * roll.spanPast() /
                        (roll.spanPast() + roll.spanFuture());
    int ahead = 0, behind = 0;
    bool rect = false;
    double rectX = 0;

    for (size_t i = 0; i < list.ops.size(); )
    {
        const int op = (int)list.ops[i++];
        int arity = cairo2d_op_arity(op);

        if (arity < 0)
            arity = (int)list.ops[i] + 2;

        if (op == CAIRO2D_RECTANGLE)
        {
            rect = true;
            rectX = list.ops[i];
        }
        else if (op == CAIRO2D_STROKE)
        {
            if (rect && rectX > nowX)
                ahead++;

            rect = false;
        }
        else if (op == CAIRO2D_FILL)
        {
            if (rect && rectX < nowX)
                behind++;

            rect = false;
        }

        i += (size_t)arity;
    }

    check(ahead > 0,
          "the scheduled future is drawn: " + std::to_string(ahead) +
          " outlined notes right of the now-line at x=" +
          std::to_string((int)nowX));

    check(behind > 0,
          "and the past is drawn: " + std::to_string(behind) +
          " filled notes left of it");

    /* ---- what a gesture does to the view ----
     *
     * The shell hands pixels in and the content decides what they mean,
     * on every platform, which is the one thing about a canvas that can
     * be wrong silently -- a drag that lands somewhere else is invisible
     * until somebody tries it. Public entry points, so ctest can try it.
     */
    const double was = roll.viewNow();

    roll.pressAt(WIDTH * 0.4, HEIGHT / 2.0, 1, 1);
    roll.motionTo(WIDTH * 0.5, HEIGHT / 2.0);
    roll.releaseAt(WIDTH * 0.5, HEIGHT / 2.0, 1);

    check(!roll.following() && roll.viewNow() < was,
          "a drag to the right scrubs back and drops out of follow");

    roll.pressAt(WIDTH * 0.5, HEIGHT / 2.0, 1, 2);

    check(roll.following(), "and a double-click goes back to live");

    roll.releaseAt(WIDTH * 0.5, HEIGHT / 2.0, 1);

    /* A motion with nothing held is a pointer crossing the roll, not a
       scrub. The browser sends one for every mouse move over the canvas
       and the desktop's motion controller does the same. */
    roll.motionTo(WIDTH * 0.1, HEIGHT / 2.0);

    check(roll.following(),
          "and a motion with nothing held moves nothing");

    /* The wheel is time. A factor above one means "draw it bigger",
       which for this canvas is less time on screen -- the one place a
       shell's number is read as something other than its name
       (RollCanvas::zoomBy). */
    const double span = roll.spanPast();

    roll.zoomBy(1.25);
    check(roll.spanPast() < span,
          "a wheel in shows less time: " + std::to_string((int)span) +
          "s -> " + std::to_string((int)roll.spanPast()) + "s");

    for (int i = 0; i < 40; i++)
        roll.zoomBy(0.8);

    check(roll.spanPast() >= 5.0 && roll.spanFuture() >= 2.5,
          "and forty notches out stop at the bounds: " +
          std::to_string((int)roll.spanPast()) + "s and " +
          std::to_string((int)roll.spanFuture()) + "s");

    /* ---- and a rewind takes the piece with it ----
     *
     * History is keyed to transport time, so notes kept across a rewind
     * sit right of the new now-line and draw as a future that already
     * happened. That is what showed up when the desktop switched pieces,
     * and it is why this class listens to sigReset at all. */
    sched.reset();

    const List after = drawRoll(roll, WIDTH, HEIGHT, 1);

    check(after.ops.size() < list.ops.size() && roll.following() &&
          roll.viewNow() == 0,
          "a rewind drops the history: " + std::to_string(list.ops.size()) +
          " words -> " + std::to_string(after.ops.size()));

    /* ---- and a bar with no end yet does not freeze the history ----
     *
     * A note delivered with duration <= 0 is live input's "held until
     * further notice", and its bar has no end until a NOTEOFF gives it
     * one. The roll used to keep those in the same queue as the ended
     * ones, where prune() walks from the front and must not drop a bar
     * that is still sounding -- so one of them at the front stopped the
     * walk, and every note behind it stayed for as long as the key was
     * down. A flushed key (thcScheduler::stop) meant forever.
     *
     * Asked of the canvas and not of the drawing, which is the one
     * thing here that cannot be read out of an op list: a bar older
     * than the window is off screen and skipped whether it was pruned
     * or not, so the leak and the fix draw the same picture.
     *
     * The note goes in through sigDelivered rather than through a key,
     * because what is being tested is the roll's own bookkeeping and
     * belfry has no `input midi' chain to press a key on. */
    roll.SetTimeSpan(5, 2.5);

    thcEvent down;

    down.type            = THC_EV_NOTE;
    down.at              = sched.now();
    down.channel         = 0;
    down.u.note.note     = 60;
    down.u.note.velocity = 100;
    down.u.note.duration = 0;          /* held: no end, ever            */

    sched.sigDelivered.emit(down);

    play(sched, synth, LONG_SECONDS);

    const double cutoff = sched.now() - 4 * roll.spanPast();

    check(roll.oldestKept() >= cutoff,
          "a bar still sounding does not freeze the history: at " +
          std::to_string((int)sched.now()) + "s the oldest kept ends at " +
          std::to_string((int)roll.oldestKept()) + "s, cutoff " +
          std::to_string((int)cutoff) + "s");

    printf("\n%s\n", failures == 0
           ? "the roll draws what it has played and what it is about to"
           : (std::to_string(failures) + " failed").c_str());

    return failures;
}

File: wasm/web/rollcheck.mjs
#!/usr/bin/env node
/*
 * Copyright (C) 2004-2026 Metaphonic Labs
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General
 * Public License along with this program; if not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*
 * rollcheck.mjs -- the piano roll, through the browser module, against the
 * desktop's.
 *
 *   node wasm/web/rollcheck.mjs [BUILD_DIR] [NATIVE_BUILD_DIR]
 *
 * The roll in a browser is the desktop's RollCanvas compiled to wasm and
 * drawn in the mirror, beside the scheduler whose pending queue is the
 * future half of what it shows. The claim that makes is that there is one
 * drawing and not two, and a claim nothing checks is how the page came to
 * have a second, smaller roll in its own sixteen colours in the first
 * place.
 *
 * So the same seeded piece is played to the same instant in both builds
 * and the roll is drawn the same number of times at the same size, and
 * the two lists of ops are compared word for word. A seeded piece replays
 * identically (docs/GEN_FORMAT.md, gencheck) and the roll's drawing is a
 * function of the piece and the frame, so anything left over is the two
 * builds disagreeing about the picture.
 *
 * The native side is scripts/rollcheck, which is linked against
 * cairo-canvas2d rather than cairo for exactly this -- its `--dump' is a
 * list of ops and not pixels. It also carries the run's own numbers, and
 * this drives itself from those: one place says how long the piece is
 * played for and how big it is drawn, rather than two that could stop
 * agreeing quietly.
 *
 * Then what drawcheck.mjs asks of every other picture, of this one: the
 * list is walkable by the arity table alone, and replay.js knows every op
 * in it, at more than one size -- because a draw takes its geometry from
 * the w and h it was handed, and a roll that cached the first is a roll
 * that is wrong in every pane but the one it opened in.
 *
 * Exit status is the number of failures.
 */

import fs from 'node:fs';
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { fileURLToPath, pathToFileURL } from 'node:url';

import { ARITY, OP_NAMES, replay } from 'cairo-canvas2d';
import { instruments } from './piececheck.mjs';
import { loadPiece } from './render.mjs';

const here = path.dirname(fileURLToPath(import.meta.url));
const top = path.join(here, '..', '..');
const build = path.resolve(process.argv[2] ?? path.join(top, 'build-web'));
const native = path.resolve(process.argv[3] ?? path.join(top, 'build'));

/* The piece. Seeded, so it replays; and one whose composers queue a phrase
   ahead rather than a note at a time, so there is a future on screen to
   have been drawn at all -- which is the half this whole roll exists for.
   scripts/rollcheck says what it does when a piece queues nothing. */
const PIECE = 'belfry.gen';

/* And the sizes the well-formedness pass draws at: a pane, and the band a
   phone in landscape leaves. */
const SIZES = [[900, 180], [360, 80]];

const BLOCK = 128;

let failures = 0;

const check = (good, what) =>
{
    process.stdout.write(`${good ? 'ok  ' : 'FAIL'}  ${what}\n`);

    if (!good)
        failures++;
};

const rollcheck = path.join(native, 'scripts', 'rollcheck');

if (!fs.existsSync(rollcheck))
{
    process.stdout.write(
        `rollcheck: no rollcheck in ${native}; build the native tree ` +
        'first -- it is the desktop half of the comparison, and it needs ' +
        "`npm ci' in wasm/web to have been run before it was configured.\n");
    process.exit(1);
}

/* ---- the desktop's list ------------------------------------------------ */

const want = JSON.parse(execFileSync(
    rollcheck,
    ['-p', path.join(native, 'plugins') + path.sep, '--dump',
     path.join(top, 'gen', PIECE)],
    { encoding: 'utf8', maxBuffer: 1 << 28,
      env: { ...process.env, THINK_DSP_PATH: path.join(top, 'dsp') } }));

/* ---- and the module's ------------------------------------------------- */

const { default: createThinkWeb } =
    await import(pathToFileURL(path.join(build, 'thinkweb.js')).href);

const dsps = instruments(build);
const gen = fs.readFileSync(path.join(build, 'gen', PIECE), 'utf8');

const { M, ok, log } = await loadPiece(createThinkWeb, {
    rate: want.rate, windowlen: want.window, block: BLOCK,
    gen, instruments: dsps,
});

if (!ok)
{
    process.stdout.write(`FAIL  ${PIECE} did not load\n      ` +
                         `${log.join('\n      ')}\n`);
    process.exit(1);
}

/* Before the transport moves, because a roll is built the moment a shell
   first says how big it is and keeps what the scheduler delivers from
   then on. One built afterwards would have an empty history and half a
   drawing, which is also what opening the pane mid-piece looks like. */
M._tw_roll_viewport(0, 0, want.width, want.height);

/* Played the way the page plays it: Play stamped for the top of the next
   window, then a block at a time, the transport stepped once per window
   inside tw_render. The native side steps to the same times by the same
   arithmetic (scripts/rollcheck, play). */
M._tw_transport(-1, 0, 0);

while (M._tw_now() < want.seconds)
{
    M._tw_render(BLOCK);
    M._tw_events_clear();
}

/* ---- the three tables, out of the heap -------------------------------- */

function readList (M)
{
    const at = M._tw_draw_ops();
    const words = M._tw_draw_words();
    const ops = words > 0
        ? M.HEAPF32.subarray(at >> 2, (at >> 2) + words)
        : new Float32Array(0);

    const strings = [];

    for (let i = 0; i < M._tw_draw_string_count(); i++)
        strings.push(M.UTF8ToString(M._tw_draw_string(i)));

    return { ops, strings };
}

/* Walked with nothing but the arity table, which is the claim the encoding
   makes. Returns the ops it counted, or a sentence. */
function walk (ops)
{
    let count = 0;

    for (let i = 0; i < ops.length; )
    {
        const op = ops[i++];

        if (OP_NAMES[op] === undefined)
            return `op ${op} at word ${i - 1} is not one`;

        let arity = ARITY[op];

        if (arity < 0)                  /* SET_DASH says its own length */
        {
            if (i >= ops.length)
                return `${OP_NAMES[op]} at word ${i - 1} has no count`;

            arity = ops[i] + 2;
        }

        if (i + arity > ops.length)
            return `${OP_NAMES[op]} at word ${i - 1} runs off the end`;

        i += arity;
        count++;
    }

    return count;
}

/* A context that draws nothing and counts what it was asked for: what is
   under test is that every op in the list is one replay.js knows, and it
   says so by throwing. drawcheck.mjs has the same stand-in. */
function counter ()
{
    const nothing = () => { ctx.calls++; };
    const ctx = {
        calls: 0,
        beginPath: nothing, moveTo: nothing, lineTo: nothing,
        bezierCurveTo: nothing, arc: nothing, rect: nothing,
        closePath: nothing, fill: nothing, stroke: nothing, clip: nothing,
        fillRect: nothing, clearRect: nothing, fillText: nothing,
        drawImage: nothing, save: nothing, restore: nothing,
        translate: nothing, scale: nothing, setTransform: nothing,
        setLineDash: nothing,
    };

    return ctx;
}

/* ---- one picture, drawn twice ----------------------------------------- */

/* The same number of frames as the native side, because the pitch range
   eases toward its fit one picture at a time: "the same frame" is the same
   count of draws as well as the same instant. */
let words = -1;

for (let i = 0; i < want.frames; i++)
    words = M._tw_roll_draw(want.width, want.height);

check(words > 0, `${PIECE} at ${want.seconds}s draws ${words} words ` +
                 `through the module`);

const got = readList(M);

if (words > 0)
{
    /* fround on the way in: the dump is %.9g, which round-trips a float
       exactly but parses here as the double nearest that decimal. The
       module's ops come out of a Float32Array and are the float itself,
       so one of the two has to be narrowed and this is the one that can
       be. */
    let first = -1;

    for (let i = 0; i < Math.max(got.ops.length, want.ops.length); i++)
        if (got.ops[i] !== Math.fround(want.ops[i]))
        {
            first = i;
            break;
        }

    check(first < 0,
          first < 0
              ? `and the desktop's ${want.ops.length} words are the same ` +
                'words, one class compiled twice'
              : `the two lists differ at word ${first}: the module says ` +
                `${got.ops[first]}, the desktop ` +
                `${Math.fround(want.ops[first])} ` +
                `(${got.ops.length} words against ${want.ops.length})`);

    const sameStrings =
        got.strings.length === want.strings.length &&
        got.strings.every((s, i) => s === want.strings[i]);

    check(sameStrings,
          sameStrings
              ? `and the ${want.strings.length} string(s) it indexes: ` +
                `${want.strings.join(', ') || 'none'}`
              : `the strings differ: [${got.strings}] against ` +
                `[${want.strings}]`);
}

/* ---- and the list is one, at more than one size ----------------------- */

for (const [w, h] of SIZES)
{
    M._tw_roll_viewport(0, 0, w, h);

    const n = M._tw_roll_draw(w, h);

    if (n <= 0)
    {
        check(false, `the roll at ${w}x${h}: drew nothing`);
        continue;
    }

    const { ops, strings } = readList(M);
    const walked = walk(ops);

    if (typeof walked === 'string')
    {
        check(false, `the roll at ${w}x${h}: ${walked}`);
        continue;
    }

    try
    {
        replay(counter(), ops, strings, [], { width: w, height: h, dpr: 1 });
        check(true, `the roll at ${w}x${h} is ${walked} ops replay.js knows`);
    }
    catch (e)
    {
        check(false, `the roll at ${w}x${h}: ${e.message}`);
    }
}

process.stdout.write(failures === 0
    ? '\nthe browser draws the desktop\'s roll, word for word\n'
    : `\n${failures} failed\n`);

process.exitCode = failures;
